src/util/recentlyEditedTree.js

```diff
@@ -0,0 +1,115 @@
+          if (timeDifference(timestamp(), descendant.lastUpdated) > editTimeMax) {
+            _.unset(tree, findClosestAncestorWithMultipleChildren(tree, ['_ROOT_', ...pathToContext(descendant.path)]).path)
+            descendantReplaced = true
+          }
+          else {
+            const updatedDescendantPath = newPath.concat(descendant.path.slice(newPath.length))
+            _.unset(tree, ['_ROOT_', ...pathToContext(descendant.path)])
+            _.set(tree, ['_ROOT_', ...pathToContext(updatedDescendantPath)], { leaf: true, lastUpdated: timestamp(), path: updatedDescendantPath })
+          }
+        })
+        if (descendantReplaced) onNodeChange(tree, oldPath, newPath) // called once again to remove merge inconsitencty that might occur while replacing descendants by ancestor
+        else _.unset(tree, oldContext)
+      }
+      else {
+        let isMerged = false // eslint-disable-line fp/no-let
+        leafNodes.forEach(descendant => {
+          const descendantContext = ['_ROOT_', ...pathToContext(descendant.path)]
+          if (descendantContext[1] === '__EM__') return // preventing thoughts at level 1 from merged to this (temporary fix)
+          const longContext = descendantContext.length > newContext.length ? descendantContext : newContext
+          const shortContext = newContext.length < descendantContext.length ? newContext : descendantContext
+          // siblings A.B.C and A.B.D ---> A.B.D
+          if (descendantContext.length === newContext.length && newContext.length - commonPath.length === 1) {
+            isMerged = true
+            _.unset(tree, descendantContext)
+            _.set(tree, newContext, { leaf: true, lastUpdated: timestamp(), path: newPath })
+          }
+          /*
+            restricting merge of direct cousins like A.B.C.F and A.B.D.E
+            merging distant relation A.B.C and A.B.D.E.F.G.H ---> A.B.D.E.F.G.H
+          */
+          else if (!(shortContext.length - commonPath.length > 1 || (newContext.length === descendantContext.length && newContext.length - commonPath.length === 2))) {
+            isMerged = true
+            _.unset(tree, shortContext.slice(0, commonPath.length + 1))
+            _.set(tree, longContext, { leaf: true, lastUpdated: timestamp(), path: newContext.length > descendantContext.length ? newPath : descendant.path })
+          }
+        })
+        // adding new thought to the tree
+        if (!isMerged) _.set(tree, newContext, { leaf: true, lastUpdated: timestamp(), path: newPath })
+      }
+    }
+  }
+}
+
+export const onNodeDelete = (tree, oldPath) => {
+  const oldContext = ['_ROOT_', ...pathToContext(oldPath)]
+  const { node: commonNode, path: commonPath } = findTreeDeepestSubcontext(tree, oldContext)
+  if (commonNode) {
+    if (oldContext.length > commonPath.length) return
+    if (commonPath.length !== 1) _.unset(tree, commonPath) // preventing deleting of _ROOT_
+    const parentPath = commonPath.slice(0, commonPath.length - 1)
```

Comment:
Let's use the existing `contextOf` method to get the `parentPath`. That offers some decoupling between the context functionality and implementation. Admittedly there are many places in the code that "know" that paths are arrays, and often need to, but it's still good to stick to the `contextOf` and `head` interfaces when appropriate to minimize this implementation-reliance.

Reply:
Okay